From 33c43ca84c9f62dd02cbd4bb6995ed237f87060f Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <periyasa@qti.qualcomm.com>
Date: Thu, 22 May 2025 16:39:13 +0530
Subject: [PATCH 06/12] UPSTREAM: DP: wifi: ath12k: Add Tx limit for data
 packets

Limit the Tx data packet per pdev based on the threshold of 0x8000
by reserving 1k for EAPOL frames. Which helps memory usage in Multiclient
setup with heavy traffic. Without the Tx limit check Multicast frames
send lead to out of memory in the system where HW capable of processing
more than the system memory profile.

Signed-off-by: Karthikeyan Periyasamy <periyasa@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/dp.h          |  6 ++++++
 drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c | 13 +++++++++++++
 drivers/net/wireless/ath/ath12k/wifi7/dp_tx.h |  1 +
 drivers/net/wireless/ath/ath12k/wifi7/hw.c    | 18 ++++++++++++++++++
 4 files changed, 38 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp.h b/drivers/net/wireless/ath/ath12k/dp.h
index 0a44d0a..7ad2b74 100644
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -218,6 +218,9 @@ struct ath12k_pdev_dp {
 #define DP_RX_BUFFER_SIZE		2048
 #endif
 
+#define ATH12K_NUM_EAPOL_RESERVE       1024
+#define ATH12K_DP_PDEV_TX_LIMIT        (ATH12K_NUM_POOL_TX_DESC - ATH12K_NUM_EAPOL_RESERVE)
+
 #define DP_WBM_RELEASE_RING_SIZE	64
 #define DP_TCL_DATA_RING_SIZE		512
 #define DP_TX_IDR_SIZE			DP_TX_COMP_RING_SIZE
@@ -465,6 +468,9 @@ struct ath12k_dp_ring_bp_stats {
 struct ath12k_device_dp_tx_err_stats {
 	/* TCL Ring Descriptor unavailable */
 	u32 desc_na[DP_TCL_NUM_RING_MAX];
+
+	u32 threshold_limit;
+
 	/* Other failures during dp_tx due to mem allocation failure
 	 * idr unavailable etc.
 	 */
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
index f232d08..ca6c690 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
@@ -70,6 +70,19 @@ static int ath12k_wifi7_dp_prepare_htt_metadata(struct sk_buff *skb)
 	return 0;
 }
 
+bool ath12k_mac_tx_check_max_limit(struct ath12k_pdev_dp *dp_pdev, struct sk_buff *skb)
+{
+	if (atomic_read(&dp_pdev->num_tx_pending) > ATH12K_DP_PDEV_TX_LIMIT) {
+		/* Allow EAPOL */
+		if (!(skb->protocol == cpu_to_be16(ETH_P_PAE))) {
+			dp_pdev->dp->device_stats.tx_err.threshold_limit++;
+			return true;
+		}
+	}
+
+	return false;
+}
+
 /* TODO: Remoe the export once this file is built with wifi7 ko */
 int ath12k_wifi7_dp_tx(struct ath12k_pdev_dp *dp_pdev,
 		       struct ath12k_link_vif *arvif,
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.h b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.h
index e644806..d1be997 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.h
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.h
@@ -14,4 +14,5 @@ int ath12k_wifi7_dp_tx(struct ath12k_pdev_dp *dp_pdev,
 		       bool is_mcast, struct ath12k_link_sta *arsta);
 u32 ath12k_wifi7_dp_tx_get_vdev_bank_config(struct ath12k_base *ab,
 					    struct ath12k_link_vif *arvif);
+bool ath12k_mac_tx_check_max_limit(struct ath12k_pdev_dp *dp_pdev, struct sk_buff *skb);
 #endif
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/hw.c b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
index 6186ed6..d5eac8b 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/hw.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
@@ -1001,6 +1001,15 @@ static void ath12k_wifi7_mac_op_tx(struct ieee80211_hw *hw,
 
 	if (!vif->valid_links || !is_mcast || is_dvlan || is_eth ||
 	    test_bit(ATH12K_FLAG_RAW_MODE, &ar->ab->dev_flags)) {
+		ret = ath12k_mac_tx_check_max_limit(dp_pdev, skb);
+		if (ret) {
+			ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
+				   "failed due to limit check pdev idx %d\n",
+				   ar->pdev_idx);
+			ieee80211_free_txskb(hw, skb);
+			return;
+		}
+
 		ret = ath12k_wifi7_dp_tx(dp_pdev, arvif, skb, false, 0, is_mcast, arsta);
 		if (unlikely(ret)) {
 			ath12k_warn(ar->ab, "failed to transmit frame %d\n", ret);
@@ -1021,6 +1030,15 @@ static void ath12k_wifi7_mac_op_tx(struct ieee80211_hw *hw,
 			tmp_dp_pdev = ath12k_dp_to_dp_pdev(tmp_ar->ab->dp, tmp_ar->pdev_idx);
 			if (!tmp_dp_pdev)
 				continue;
+
+			ret = ath12k_mac_tx_check_max_limit(tmp_dp_pdev, skb);
+			if (ret) {
+				ath12k_dbg(tmp_ar->ab, ATH12K_DBG_MAC,
+					   "failed mcast tx due to limit check pdev idx %d\n",
+					    tmp_ar->pdev_idx);
+				continue;
+			}
+
 			msdu_copied = skb_copy(skb, GFP_ATOMIC);
 			if (!msdu_copied) {
 				ath12k_err(ar->ab,
-- 
2.34.1

